deepinv/datasets/fastmri.py

```diff
@@ -0,0 +1,248 @@
+r"""Pytorch Dataset for fastMRI.
+
+Code modified from: https://github.com/facebookresearch/fastMRI/blob/main/fastmri/data/mri_data.py
+
+Copyright (c) Facebook, Inc. and its affiliates.
+
+This source code is licensed under the MIT license found in the
+LICENSE file in the root directory of this source tree.
+"""
+
+import random
+from pathlib import Path
+from typing import (
+    Any,
+    Callable,
+    NamedTuple,
+    Optional,
+    Union,
+)
+import os
+import h5py
+import torch
+import numpy as np
+
+
+class SliceSampleFileIdentifier(NamedTuple):
+    fname: Path
+    slice_ind: int
+
+
+class FastMRISliceDataset(torch.utils.data.Dataset):
+    """Dataset for `fastMRI <https://fastmri.med.nyu.edu/>`_ that provides access to MR image slices.
+
+    | 1) The fastMRI dataset includes two types of MRI scans: knee MRIs and
+    | the brain (neuro) MRIs, and containing training, validation, and masked test sets.
+    | 2) MRIs are volumes (3D) made of slices (2D).
+    | 3) This class in particular considers one data sample as one slice of a MRI scan,
+    | thus slices of the same MRI scan are considered independently in the dataset.
+
+
+    **Raw data file structure:** ::
+
+        self.root --- file1000005.h5
+                   |
+                   -- xxxxxxxxxxx.h5
+
+    | 0) To download raw data, please go to the bottom of the page `https://fastmri.med.nyu.edu/`
+    | 1) Each MRI scan is stored in a HDF5 file and can be read with the h5py package.
+    | Each file contains the k-space data, ground truth and some meta data related to the scan.
+    | 2) MRI scans can either be single-coil or multi-coil with each coil in
+    | a multi-coil MRI scan focusses on a different region of the image.
+    | 3) In multi-coil MRIs, k-space has the following shape:
+    | (number of slices, number of coils, height, width)
+    | 4) For single-coil MRIs, k-space has the following shape:
+    | (number of slices, height, width)
+
+    :param Union[str, Path] root: Path to the dataset.
+    :param bool test: Whether the split is the "test" set.
+    :param str challenge: "singlecoil" or "multicoil" depending on the type of mri scan.
+    :param bool load_metadata_from_cache: Whether to load dataset metadata from cache.
+    :param bool save_metadata_to_cache: Whether to cache dataset metadata.
+    :param Union[str, Path] metadata_cache_file: A file used to cache dataset
+        information for faster load times.
+    :param callable, optional sample_filter: A callable object that takes a
+        SliceSampleFileIdentifier as input and returns a boolean indicating
```

Comment:
add a ref to the class with :meth:`deepinv.datasets.SliceSampleFileIdentifier` (check that this works)